File: 739-daily-temperatures/daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];
        Stack<int[]> stack = new Stack<>();

        for (int i = 0; i < temperatures.length; i++) {
            int temp = temperatures[i];
            while (!stack.isEmpty() && temp > stack.peek()[0]) {
                int[] top = stack.pop();
                int index = top[1];
                res[index] = i - index;
            }
            stack.push(new int[]{temp, i});
        }        
        return res;
    }
}

File: 739-daily-temperatures/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # time: O(n)
        # space: O(n)
        res = [0] * len(temperatures)
        stack = []

        for i, temp in enumerate(temperatures):
            while stack and temp > stack[-1][1]:
                prev_i, prev_temp = stack.pop()  
                res[prev_i] = i - prev_i
            stack.append([i, temp])
        
        return res
